### PR TITLE
Hotel model endpoints

### PR DESCRIPTION
**Pull request by mark-marshall**
# Description

Please include a summary of the change and a link to which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

The three company endpoints [POST], [GET], and [PUT] with documentation. These endpoints require formal testing once agreed upon and also highlight some other areas up for discussion:

(1) This work includes input validation and formatting middleware - validation may be covered off by Jooigoose as discussed. Middleware folder structure and general approach need to be discussed also.

(2) There are multiple syntax options to achieve the same thing with Mongoose/MongoDb. We need to agree on which syntax to use for consistency.

Fixes # (issue)

#23 

## Change status

- [x] Complete, but not tested (may need new tests)

# How Has This Been Tested?
This has been tested with Postman but needs to be formally tested with jest/other once an approach to this is agreed on.

Please describe steps taken to ensure this feature is functional and does not break other features:
Altered code to go back to working version, checked for log errors.

# Checklist:

- [x] My code follows the style guidelines of this project

- [x] I have performed a self-review of my own code

- [x] I have commented my code, particularly in hard-to-understand areas

- [x] I have made corresponding changes to the documentation, if relevan

- [x] My changes generate no new warnings

- [x] There are no merge conflicts

### CODE REVIEW

File: middleware/formatHotelPost.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const formatHotelPost = (req, res, next) => {
+  const hotel = req.body;
+  hotel.name = hotel.name
+    .toLowerCase()
+    .replace(/\b([a-z])/gi, char => char.toUpperCase());
+  req.body = { ...hotel };
+  next();
+};
+
+module.exports = formatHotelPost;
```
Comment (Pav0l) [inline middleware/formatHotelPost.js]: Isn't this module doing the same thing as `middleware/formatHotelChange.js` ?

File: middleware/validateHotelChange.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    res.status(400).json(error.invalidHotelPost);
+  }
+  else if (hotel.rooms) {
+    res.status(400).json(error.restrictedHotelPut);
+  } else {
+    next();
+  }
+};
+
+module.exports = validateHotelChange;
```
Comment (Pav0l) [inline middleware/validateHotelChange.js:15]: ❤️ this validation

File: models/index.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 const mongoose = require('mongoose');
-const User = require('./usersModel');
+
+const User = require('./usersSchema');
```
Comment (Pav0l) [inline models/index.js]: I believe this should stay `./usersModel` to be consistent with other models file naming convention

File: routes/hotelRoutes.js [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+ */
+routes.post('/', validateHotelPost, formatHotelPost, async (req, res, next) => {
+  const hotel = req.body;
+  try {
+    const newHotel = new models.Hotel(hotel);
+    await newHotel.save();
+    if (newHotel._id) {
+      res.status(201).json(newHotel);
+    } else {
+      res.status(400).json(error.addHotel);
```
Comment (Pav0l) [inline routes/hotelRoutes.js]: Will this also returns error message, if the `hotel` already exist in DB?
If not, please add this check.

**Review by Pav0l — CHANGES_REQUESTED** [review]
Awesome job with documentation and suggestions! 🥇 

Please check review comments in code.

File: routes/hotelRoutes.js (hunk 4 above)
Comment (mark-marshall) [reply routes/hotelRoutes.js]: Nice catch. I have added this now but it is only by name - cases aligned to how they will be formatted in the db.

File: middleware/formatHotelPost.js (hunk 1 above)
Comment (mark-marshall) [reply middleware/formatHotelPost.js]: The slight diff was that formatHotelChange first had a check to see whether req.body included a name change. As this conditional will pass every time for formatHotelPost, I just combined the two together now.

File: middleware/validateHotelPost.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const { models } = require('../models/index');
+const error = require('../utils/error');
+
+const validateHotelPost = async (req, res, next) => {
+  const hotel = req.body;
+  const hotelName = hotel.name
+    .toLowerCase()
+    .replace(/\b([a-z])/gi, char => char.toUpperCase());
```
Comment (Pav0l) [inline middleware/validateHotelPost.js]: This can be removed and replaced by `middleware/formatHotel.js`

File: middleware/validateObjectId.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+const validateObjectId = (req, res, next) => {
+  const { id } = req.params;
+  if (id.match(/^[0-9a-fA-F]{24}$/)) {
+    next();
+  } else {
+    res.status(400).json(error.invalidObjectId);
+  }
+};
+
+module.exports = validateObjectId;
```
Comment (Pav0l) [inline middleware/validateObjectId.js:12]: Awesome middleware!!!

File: routes/hotelRoutes.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+      res.status(400).json(error.noHotel);
+    }
+  } catch (error) {
+    next(error);
+  }
+});
+
+/**
+ *  @api {get} api/hotel/:id Put hotel information
+ *  @apiVersion 0.1.0
```
Comment (Pav0l) [inline routes/hotelRoutes.js:138]: put instead of { get }

File: routes/hotelRoutes.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+ *      "message": "message": "a hotel must be added with at least a name and motto"
+ *    }
+ *  @apiErrorExample Error-Response: changes requested that are the same as  existing hotel information
+ *    HTTP/1.1 400 BAD REQUEST
+ *    {
+ *      "message": "the hotel could not be updated or you did not provide any new information"
+ *    }
+ */
+routes.put('/:id', validateHotelChange, formatHotel, async (req, res, next) => {
+  const { id } = req.params;
```
Comment (Pav0l) [inline routes/hotelRoutes.js]: needs the `validateObjectId` middleware

File: routes/hotelRoutes.js [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+  const hotelUpdates = req.body;
+  try {
+    const updateResult = await models.Hotel.updateOne(
+      { _id: id },
+      hotelUpdates,
+    );
+    const updatedHotel = await models.Hotel.findById(id);
+    if (updateResult.nModified) {
```
Comment (Pav0l) [inline routes/hotelRoutes.js]: check if hotel with `/:id` exists first, then update it

File: routes/hotelRoutes.js (hunk 8 above)
Comment (Pav0l) [reply routes/hotelRoutes.js]: test req.params with `/:_id`

**Review by Pav0l — CHANGES_REQUESTED** [review]
Awesome job man!

File: utils/documentExists.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+      return true;
+    } else {
+      return false;
+    }
+  } catch (error) {
+    console.error(error);
+  }
+};
+
+module.exports = documentExists;
```
Comment (Pav0l) [inline utils/documentExists.js:16]: Great utility function. Will definitely reuse it!

**Review by Pav0l — APPROVED** [review]
As always, great job!

File: utils/documentExists.js (hunk 10 above)
Comment (Velsu) [reply utils/documentExists.js]: This is indeed interesting.

**Review by Velsu — COMMENTED** [review]
Nice job 👍

File: utils/documentExists.js (hunk 10 above)
Comment (Velsu) [reply utils/documentExists.js:16]: This is indeed interesting 🔥

**Pull request merged**